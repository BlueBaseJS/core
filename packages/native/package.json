{
  "name": "bluerain-boilerplate-native",
  "version": "0.3.0",
  "description": "Hello BlueRain!",
  "author": "BlueEast Team <code@blueeast.com>",
  "private": true,
  "main": "node_modules/expo/AppEntry.js",
  "scripts": {
    "start": "exp start"
  },
  "dependencies": {
    "@blueeast/bluerain-os": "git://github.com/BlueEastCode/bluerain-os.git#feature/custom_platform",
    "expo": "19.0.1",
    "react": "16.0.0-alpha.12",
    "react-dom": "^16.0.0-alpha.12",
    "react-native": "0.46.4",
    "react-native-windows": "^0.33.0",
    "reactxp": "^0.42.0-rc.25",
    "reactxp-navigation": "1.0.5",
    "reactxp-video": "^0.2.0"
  },
  "devDependencies": {
    "@types/node": "^7.0.12",
    "@types/react": "15.0.39",
    "@types/react-native": "0.46.7",
    "@types/webpack": "^2.2.14",
    "source-map-loader": "^0.1.6"
  }
}
