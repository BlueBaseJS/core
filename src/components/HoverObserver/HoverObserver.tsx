import { HoverObserverProps } from '@bluebase/components';

import { renderChildrenWithProps } from '../../utils';

export interface HoverObserverState {
	readonly isHovering: boolean;
}

export const HoverObserver = (props: HoverObserverProps) => {
	return renderChildrenWithProps(props.children, { isHovering: false });
};

HoverObserver.displayName = 'HoverObserver';
