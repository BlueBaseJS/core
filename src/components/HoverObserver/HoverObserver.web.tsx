import React, { SyntheticEvent } from 'react';
import { HoverObserverProps } from '@bluebase/components';
import { renderChildrenWithProps } from '../../utils';

export interface HoverObserverState {
	readonly isHovering: boolean;
}

const Noop = () => { return; };

/**
 * # ðŸ›¸ HoverObserver
 *
 * A React component that notifies its children of hover interactions.
 *
 * Initial code taken from: https://github.com/ethanselzer/react-hover-observer.
 *
 * ## Usage
 * ```jsx
 * <HoverObserver>
 *  {({ isHovering }) => (
 * 		 <YourChildComponent isActive={isHovering} />
 *  )}
 * </HoverObserver>
 * ```
 */
export class HoverObserver extends React.PureComponent<HoverObserverProps, HoverObserverState> {

	public static defaultProps: Partial<HoverObserverProps> = {
		hoverDelayInMs: 0,
		hoverOffDelayInMs: 0,
		onHoverChanged: Noop,
		onMouseEnter: ({ setIsHovering }) => setIsHovering(),
		onMouseLeave: ({ unsetIsHovering }) => unsetIsHovering(),
		onMouseOut: Noop,
		onMouseOver: Noop,
	};

	private timerIds: number[] = [];

	constructor(props: HoverObserverProps) {
		super(props);

		this.state = {
			isHovering: false
		};

		this.onMouseEnter = this.onMouseEnter.bind(this);
		this.onMouseLeave = this.onMouseLeave.bind(this);
		this.onMouseOver = this.onMouseOver.bind(this);
		this.onMouseOut = this.onMouseOut.bind(this);
		this.setIsHovering = this.setIsHovering.bind(this);
		this.unsetIsHovering = this.unsetIsHovering.bind(this);
		this.componentWillUnmount = this.componentWillUnmount.bind(this);

		this.timerIds = [];
	}

	onMouseEnter(e: SyntheticEvent) {
		this.props.onMouseEnter!({
			e,
			setIsHovering: this.setIsHovering,
			unsetIsHovering: this.unsetIsHovering
		});
	}

	onMouseLeave(e: SyntheticEvent) {
		this.props.onMouseLeave!({
			e,
			setIsHovering: this.setIsHovering,
			unsetIsHovering: this.unsetIsHovering
		});
	}

	onMouseOver(e: SyntheticEvent) {
		this.props.onMouseOver!({
			e,
			setIsHovering: this.setIsHovering,
			unsetIsHovering: this.unsetIsHovering
		});
	}

	onMouseOut(e: SyntheticEvent) {
		this.props.onMouseOut!({
			e,
			setIsHovering: this.setIsHovering,
			unsetIsHovering: this.unsetIsHovering
		});
	}

	componentWillUnmount() {
		this.clearTimers();
	}

	setIsHovering() {
		this.clearTimers();

		const hoverScheduleId = setTimeout(() => {
			const newState = { isHovering: true };
			this.setState(newState, () => {
				this.props.onHoverChanged!(newState);
			});
		}, this.props.hoverDelayInMs) as any;

		this.timerIds.push(hoverScheduleId);
	}

	unsetIsHovering() {
		this.clearTimers();

		const hoverOffScheduleId = setTimeout(() => {
			const newState = { isHovering: false };
			this.setState(newState, () => {
				this.props.onHoverChanged!(newState);
			});
		}, this.props.hoverOffDelayInMs) as any;

		this.timerIds.push(hoverOffScheduleId);
	}

	clearTimers() {
		const ids = this.timerIds;
		while (ids.length) {
			const id = ids.pop() as number;
			clearTimeout(id);
		}
	}

	render() {
		const { children } = this.props;

		return (
			<div {...{
				onMouseEnter: this.onMouseEnter,
				onMouseLeave: this.onMouseLeave,
				onMouseOut: this.onMouseOut,
				onMouseOver: this.onMouseOver,
			}}>
				{renderChildrenWithProps(children, this.state)}
			</div>
		);
	}
}