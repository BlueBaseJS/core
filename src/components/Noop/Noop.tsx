import React from 'react';

export interface NoopProps {
	children?: React.ReactNode,
}

/**
 * # 😐 Noop
 *
 * A component that does... nothing!
 */
export const Noop: React.FunctionComponent<NoopProps> = ({ children }) => {

	if (children && React.isValidElement(children)) {
		return children;
	}

	return null;
};
