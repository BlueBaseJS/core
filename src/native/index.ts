export * from './ActivityIndicator';
export * from './Button';
export * from './Image';
export * from './Text';
export * from './View';
