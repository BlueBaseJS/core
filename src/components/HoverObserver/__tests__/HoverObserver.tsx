import * as Native from '../../../native';
import { Noop } from '../../Noop';
import { HoverObserver } from '../HoverObserver';
import React from 'react';
import TestRenderer from 'react-test-renderer';

// beforeEach(() => {
// 	jest.resetModules();
// });


describe('HoverObserver', () => {

	test(`Snapshot HoverObserver component`, () => {
		const component = TestRenderer.create(
			<HoverObserver />
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	test(`Snapshot HoverObserver component with child`, () => {
		const component = TestRenderer.create(
			<HoverObserver>
				<Native.Text>Hello</Native.Text>
			</HoverObserver>
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	test(`Snapshot HoverObserver component with child as function`, () => {
		const component = TestRenderer.create(
			<HoverObserver>
				{() => (
					<Native.Text>Hello</Native.Text>
				)}
			</HoverObserver>
		);
		const tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	test(`Snapshot HoverObserver component with onMouse actions without setTimeout`, () => {
		const component = TestRenderer.create(
			<HoverObserver>
				<Native.Text>Hello</Native.Text>
			</HoverObserver>
		);
		component.root.findByType('div').props.onMouseEnter();
		expect(component.toJSON()).toMatchSnapshot();
		component.root.findByType('div').props.onMouseLeave();
		expect(component.toJSON()).toMatchSnapshot();
		component.root.findByType('div').props.onMouseOver();
		expect(component.toJSON()).toMatchSnapshot();
		component.root.findByType('div').props.onMouseOut();
		expect(component.toJSON()).toMatchSnapshot();
	});

	test(`Snapshot HoverObserver component with onMouse actions after setTimeout`, (done) => {
		const component = TestRenderer.create(
			<HoverObserver>
				<Native.Text>Hello</Native.Text>
			</HoverObserver>
		);
		setTimeout(() => {
			component.root.findByType('div').props.onMouseEnter();
			expect(component.toJSON()).toMatchSnapshot();
			component.root.findByType('div').props.onMouseLeave();
			expect(component.toJSON()).toMatchSnapshot();
			component.root.findByType('div').props.onMouseOver();
			expect(component.toJSON()).toMatchSnapshot();
			component.root.findByType('div').props.onMouseOut();
			expect(component.toJSON()).toMatchSnapshot();
			done();
		});
	});

	test(`Snapshot HoverObserver component with onMouseEnter`, (done) => {
		const component = TestRenderer.create(
			<HoverObserver
				onMouseEnter={(state) => { state.setIsHovering(); }}
			>
				<Native.Text>Hello</Native.Text>
			</HoverObserver>
		);
		component.root.findByType('div').props.onMouseEnter();
		setTimeout(() => {
			expect(component.toJSON()).toMatchSnapshot();
			done();
		});
	});

	test(`Snapshot HoverObserver component with onMouse actions after setTimeout with props`, (done) => {
		const component = TestRenderer.create(
			<HoverObserver
				onHoverChanged={Noop}
				onMouseEnter={Noop}
				onMouseLeave={Noop}
				onMouseOut={Noop}
				onMouseOver={Noop}
				hoverDelayInMs={100}
				hoverOffDelayInMs={100}
			>
				<Native.Text>Hello</Native.Text>
			</HoverObserver>
		);
		const instance: any = component.getInstance();
		setTimeout(() => {
			component.root.findByType('div').props.onMouseEnter();
			expect(component.toJSON()).toMatchSnapshot();
			component.root.findByType('div').props.onMouseLeave();
			expect(component.toJSON()).toMatchSnapshot();
			component.root.findByType('div').props.onMouseOver();
			expect(component.toJSON()).toMatchSnapshot();
			component.root.findByType('div').props.onMouseOut();
			expect(component.toJSON()).toMatchSnapshot();
			instance.componentWillUnmount();
			expect(component.toJSON()).toMatchSnapshot();

			done();
		}, 200);
	});

});
