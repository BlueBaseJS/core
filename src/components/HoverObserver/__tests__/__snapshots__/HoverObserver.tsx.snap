// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoverObserver Snapshot HoverObserver component 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
/>
`;

exports[`HoverObserver Snapshot HoverObserver component with child 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with child as function 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouse actions after setTimeout 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouse actions after setTimeout 2`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouse actions after setTimeout 3`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouse actions after setTimeout 4`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouse actions after setTimeout with props 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouse actions after setTimeout with props 2`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouse actions after setTimeout with props 3`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouse actions after setTimeout with props 4`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouse actions after setTimeout with props 5`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouse actions without setTimeout 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouse actions without setTimeout 2`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouse actions without setTimeout 3`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouse actions without setTimeout 4`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouseEnter 1`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouseEnter 2`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouseEnter 3`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;

exports[`HoverObserver Snapshot HoverObserver component with onMouseEnter 4`] = `
<div
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <Text
    style={
      Array [
        undefined,
        undefined,
      ]
    }
  >
    Hello
  </Text>
</div>
`;
